### PR TITLE
Properly Remove Expired Robots In Network Input

### PR DESCRIPTION
**Pull request by garethellis0**
### Description
Properly removed expired robots using the latest timestamp out of all received camera images. 

Still needs some tweaks from @MathewMacDougall.

### Testing Done
Ran `network_input` and it published correct worlds from grsim.

### Resolved Issues
NA

### Length Justification
NA

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/network_input/networking/network_client.cpp [hunk 1]
```diff
@@ -90,12 +90,20 @@ void NetworkClient::filterAndPublishVisionData(SSL_WrapperPacket packet)
             ret.first->second = detection;
         }
 
+        std::optional<Timestamp> latest_timestamp = std::nullopt;
```
Comment (EvanMorcom) [inline src/thunderbots/software/network_input/networking/network_client.cpp]: I think latest_detections could be confused as either 'the newest' or 'a delayed/old detections'.

I think what you're going for is 'most_recent/newest_detections"?

File: src/thunderbots/software/network_input/networking/network_client.cpp [hunk 2]
```diff
@@ -90,12 +90,20 @@ void NetworkClient::filterAndPublishVisionData(SSL_WrapperPacket packet)
...
         // Create a vector of the latest detection data for each camera
         std::vector<SSL_DetectionFrame> latest_detections;
         for (auto it = latest_detection_data.begin(); it != latest_detection_data.end();
              it++)
         {
             latest_detections.push_back(it->second);
+            Timestamp timestamp = Timestamp::fromSeconds(it->second.t_capture());
+
+            // Remember the latest timestamp
+            if (!latest_timestamp || timestamp > latest_timestamp)
```
Comment (EvanMorcom) [inline src/thunderbots/software/network_input/networking/network_client.cpp]: Nit:
I like seeing "!latest_timestamp.has_value()", I feel like it's easier to understand the datatype at a glance and what you're checking.
Up to you though.

**Review by EvanMorcom — COMMENTED** [review]
LGTM 💯 
Couple comments

**Review by MathewMacDougall — COMMENTED** [review]
Nothing to add beyond Evan's comments

**Comment by garethellis0** [comment]
Sorry, I should've been more clear initially. These are @MathewMacDougall's changes. @MathewMacDougall - could you handle this one, since I'm really not familiar with this part of the codebase?

File: src/thunderbots/software/network_input/networking/network_client.cpp (hunk 1 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/networking/network_client.cpp]: Ultimately removed this variable so this is resolved

**Review by MathewMacDougall — APPROVED** [review]
LGTM

**Comment by henryymliu** [comment]
Given the issues seen with the demo, is it appropriate to revert the changes by this PR and reopen the issue?

**Comment by MathewMacDougall** [comment]
Was this what you reverted? I had unmerged filter changes in the branch I gave you, which is what I thought you were referring to.

**Comment by henryymliu** [comment]
Yes, this was the commit I reverted, and stuff worked afterwards.

**Pull request merged**